File: code/models/volfctw.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <memory>
#include <numeric>
#include <utility>
#include <vector>
#include <iomanip>
#include "corpus.hpp"
#include "utils.hpp"


//template <typename> struct Typedebug;
template <std::size_t num_children>
class VolfNode {
    using idx_t = size_t;
    using IdxContext = Context<idx_t>;
    using ProbAr = std::array<double, num_children>;
    using count_t = uint8_t;
    // using count_t = std::size_t;
    std::array<std::unique_ptr<VolfNode>, num_children> m_children{};
    std::array<count_t, num_children> m_counts{};
    std::size_t m_total{};
    double m_beta{1.0};
    double m_alpha;
    std::unique_ptr<VolfNode> make_child() {
        return std::make_unique<VolfNode>(m_alpha);
    }
    VolfNode& get_child(idx_t idx) {
        if (!m_children[idx]) {
            m_children[idx] = make_child();
        }
        return *m_children[idx];
    }
    VolfNode const& get_child(idx_t idx) const {
        assert(m_children[idx]);
        return *m_children[idx];
    }
    ProbAr get_theoretical_new_child_pe_ar() const {
        ProbAr tmp;
        std::fill(tmp.begin(), tmp.end(), 1.0/num_children);
        return tmp;
    }
    ProbAr get_pe_ar() const {
        ProbAr tmp;
        std::transform(m_counts.begin(), m_counts.end(), tmp.begin(),
                       [this](auto const& el) {
                           return (el + (1/m_alpha))
                               / (m_total + (num_children/m_alpha));
                       });
        return tmp;
    }
    void theoretical_apply_weighting(ProbAr &prob_ar) const {
        auto sum = 0.0;
        std::transform(prob_ar.begin(), prob_ar.end(), prob_ar.begin(),
                       [this,&sum] (auto const &prob) {
                           auto val = prob + (1.0/num_children);
                           sum += val;
                           return val;
                       });
        std::transform(prob_ar.begin(), prob_ar.end(), prob_ar.begin(),
                       [sum] (auto const &prob) {
                           return prob/sum;
                       });
    }
    void apply_weighting(ProbAr &prob_ar) const {
        auto sum = 0.0;
        std::transform(prob_ar.begin(), prob_ar.end(), m_counts.begin(), prob_ar.begin(),
                       [&sum, this](auto const& prob, auto const& count) {
                           auto val =prob + (count + 1/m_alpha)*beta_tag();
                           sum += val;
                           return val;
                       });
        // auto sum = std::accumulate(prob_ar.begin(), prob_ar.end(), 0.0);
        std::transform(prob_ar.begin(), prob_ar.end(), prob_ar.begin(),
                       [sum](auto const &prob) {
                           return prob/sum;
                       });
    }

    auto beta_tag() const {
        return m_beta / (num_children/m_alpha + m_total);
    }
    void update_beta(count_t count, double child_pw) {
        // Update beta, consts from thesis
        auto const beta_thresh = 1500000;
        if ((m_beta > beta_thresh) || (m_beta < (1.0/beta_thresh))) {
            m_beta /= 2;
        } else {
            m_beta = (count + (1/m_alpha))*beta_tag()/child_pw;
        }
    }
    void update_counts(idx_t sym) {
        m_counts[sym] += 1;
        m_total += 1;
        auto const RESCALE_THRESHOLD = std::numeric_limits<uint8_t>::max();
        if (m_counts[sym] == RESCALE_THRESHOLD) {
            auto new_total = 0;
            std::transform(m_counts.begin(), m_counts.end(), m_counts.begin(),
                           [this, &new_total](auto const &el) {
                               // Divide by 2, round up
                               auto nc =(el >> 1) + (el & 1);
                               new_total += nc;
                               return nc;
                           });
            m_total = new_total;

        }
    }
public:
    static std::size_t num_created;
    VolfNode(double alpha) : m_alpha{alpha} {++num_created;}
    ProbAr get_probs(IdxContext const &ctx) const {
        if (!ctx) {
            return get_pe_ar();
        }
        auto idx = ctx.back();
        ProbAr child_probs;
        if (m_children[idx]) {
            auto &child = get_child(idx);
            child_probs = child.get_probs(ctx.popped());
        } else {
            // Unroll children to maintain const-ness
            child_probs = get_theoretical_new_child_pe_ar();
            for (auto ictx = ctx; ictx; ictx=ictx.popped()) {
                theoretical_apply_weighting(child_probs);
            }
        }
        apply_weighting(child_probs);
        return child_probs;
    }
    ProbAr learn(IdxContext const& ctx, idx_t sym) {
        ProbAr pv;
        if (ctx) {
            auto idx = ctx.back();
            auto &child = get_child(idx);
            pv = child.learn(ctx.popped(), sym);
            auto child_pw = pv[sym];
            apply_weighting(pv);
            update_beta(m_counts[sym], child_pw);
        } else {
            pv = get_pe_ar();
        }
        update_counts(sym);
        return pv;
    }
};
// Give space for statics here.
template<std::size_t N>
std::size_t VolfNode<N>::num_created;


template <typename Alphabet>
class VolfModel {
    // Standard K-Ary implementation discussed as "Solution 1"
    using idx_t = std::size_t;

    MemoryDeque<idx_t> m_past_idxs;
    VolfNode<Alphabet::size> m_root;
    auto get_probs() const{
        return m_root.get_probs(m_past_idxs.view());
    }
public:
    using sym_t = typename Alphabet::dtype;
    VolfModel(size_t depth, double alpha)
        : m_past_idxs{depth}
        , m_root{alpha}
        {
            VolfNode<Alphabet::size>::num_created = 0;
        }
    Footprint footprint() const {
        return {.num_nodes = VolfNode<Alphabet::size>::num_created,
                .node_size = sizeof(VolfNode<Alphabet::size>)};
    }
    void learn(sym_t sym) {
        auto idx = Alphabet::to_idx(sym);
        auto view = m_past_idxs.view();
        m_root.learn(view, idx);
        m_past_idxs.push_back(idx);
    }
    double pmf(sym_t sym) const {
        auto idx = Alphabet::to_idx(sym);
        return get_probs()[idx];
    }
    double excmf(sym_t sym) const {
        auto probar = get_probs();
        auto cmf = std::accumulate(probar.begin(),
                                   probar.begin()+Alphabet::to_idx(sym),
                                   0.0);
        return cmf;
    }
    sym_t find_sym_from_cum_prob(double cum_prob) const {
        auto probar = get_probs();
        auto pos = std::find_if(probar.begin(), probar.end(),
                                [&cum_prob](auto const& prob) {
                                    if (cum_prob < prob) {
                                        return true;
                                    }
                                    cum_prob -= prob;
                                    return false;
                                });
        return Alphabet::to_sym(std::distance(probar.begin(), pos));
    }
};

#include <sys/resource.h>
#include "ac.hpp"
int main() {
    // 3GB limit
    auto tgb = (1ul<<30)*3;
    struct rlimit limit{tgb, tgb};
    if (-1 == setrlimit(RLIMIT_DATA, &limit)) {
        std::cerr << "Failed to set mem safety" << "\n";
        return -1;
    }
    auto dmax{9};
    correctness_and_entropy_test([]() {
        return VolfModel<BitAlphabet>(10, 15.0);
    });
    // for (auto const &name: calgary_names) {
    //     auto contents = load_file_in_memory(calgary_name_to_path.at(name));
    //     std::cout << name << " ";
    // }
    // std::cout << "File,Alphabet,Depth,Entropy" << std::endl;
    // for (auto const & name: calgary_names) {
    //     auto const path = calgary_name_to_path.at(name);
    //     auto contents = load_file_in_memory(path);
    //     for (int depth = 1; depth < dmax; ++depth) {
    //         std::cout << name << ",Bit," << depth << ","
    //                   << std::setprecision(7)
    //                   << entropy_of_model(contents.bits, VolfModel<BitAlphabet>(depth, 15.0))
    //                   << std::endl;
    //     }
    //     for (int depth = 1; depth < dmax; ++depth) {
    //         std::cout << name << ",Byte," << depth << ","
    //                   << std::setprecision(7)
    //                   << entropy_of_model(contents.bytes, VolfModel<ByteAlphabet>(depth, 15.0))
    //                   << std::endl;
    //     }
    // }

}

File: code/models/volfctw.py
# Adapted from https://github.com/omerktz/VMMPredictor/blob/master/vmms/vmm/vmm/algs/ctw/VolfNode.java
import numpy as np
RESCALE_THRESHOLD_COUNT = 255
DEFAULT_ALPHA=16
class VolfNode:
    def __init__(self, syms, alpha_factor):
        self.syms = syms
        self.counts = np.zeros(len(syms),dtype=np.int32)#[0]*len(syms)
        self.children = [None]*len(syms)
        self.alpha = alpha_factor
        self.alphainv = 1/self.alpha
        self.absizedivalpha = len(syms)/self.alpha
        self.beta = 1.0
        self.i = 0
    def _make_child(self):
        return VolfNode(self.syms, self.alpha)
    def get_child(self, idx):
        if self.children[idx] is None:
            self.children[idx] = self._make_child()
        return self.children[idx]
    def get_probs(self, ctx):
        if len(ctx):
            lc = ctx[-1]
            child = self.get_child(lc)
            child_probs = child.get_probs(ctx[:-1])
            cond_probs = self._weight_probs(child_probs)
            return cond_probs
        return self._pe_arr()
    def _weight_probs(self, probs):
        beta_tag = self.beta / (self.absizedivalpha + self.counts.sum())
        child_probs = probs + (self.counts+self.alphainv)*beta_tag
        return child_probs/child_probs.sum()
    def _pe_arr(self):
        return (self.counts +self.alphainv)/(self.counts.sum() + self.absizedivalpha)
    def _incr_count(self, sym):
        self.counts[sym] += 1
        # Rescale if needed
        if self.counts[sym] < RESCALE_THRESHOLD_COUNT:
            return
        # div by 2, and round up
        self.counts = (self.counts >> 1) + (self.counts&1)

    def learn(self, ctx, sym):
        if len(ctx):
            lc = ctx[-1]
            child = self.get_child(lc)
            child_probs = child.learn(ctx[:-1], sym)
            cond_probs = self._weight_probs(child_probs)
            # update beta, consts are from thesis
            if (self.beta > 1500000) or (self.beta < (1.0/1500000)):
                self.beta /= 2
            else:
                child_pw = child_probs[sym]
                beta_tag = self.beta / (self.counts.sum() + self.absizedivalpha)
                self.beta = (beta_tag * (self.counts[sym] + self.alphainv))/child_pw
            ret = cond_probs
        else:
            ret = self._pe_arr()
        self._incr_count(sym)
        return ret

    def prob(self, ctx, sym):
        probarr = self.get_probs(ctx)
        return probarr[sym]

from ctw import MemoryDeque
from itertools import accumulate
class VolfModel:
    def __init__(self, syms, depth, alpha=DEFAULT_ALPHA):
        self.syms = syms
        self.sym_to_idx = {s:i for i,s in enumerate(syms)}
        self.idx_to_sym = {i:s for i,s in enumerate(syms)}
        self.root = VolfNode(list(range(len(syms))), alpha)
        self.past_idxs = MemoryDeque(maxlen=depth)
    def learn(self, sym):
        idx = self.sym_to_idx[sym]
        self.root.learn(ctx=self.past_idxs.view(), sym=idx)
        self.past_idxs.append(idx)
    def probs(self):
        return self.root.get_probs(self.past_idxs.view())
    def pmf(self, sym):
        return self.probs()[self.sym_to_idx[sym]]
    def exclusive_cmf(self, sym):
        return sum(self.probs()[:self.sym_to_idx[sym]])
    def find_sym_from_cum_prob(self, cum_prob):
        probs = self.probs()
        cdfs = list(accumulate(probs, initial=0))
        ranges = [(a,b) for a,b in zip(cdfs, cdfs[1:])]
        idx = [i for i, (a,b) in enumerate(ranges) if cum_prob >= a and cum_prob < b][0]
        return self.idx_to_sym[idx]


import sys
sys.path.insert(0,"../")
from corpus.corpus import get_sum, bit_iter, byte_iter
from scripts.order0test import ac_correctness, entropy_stats

# ctor = lambda syms: VolfModel(syms=syms, depth=2)
# ac_correctness(ctor)
# entropy_stats(ctor)
# sum_text = get_sum()
bits = list(bit_iter(get_sum()))

# vm = VolfModel([0,1], 5)
from tqdm import tqdm
vm = VolfModel([0, 1], 5)
for b in tqdm(bits):
    vm.learn(b)
    #print(f"Learned: {b} pmf: {vm.pmf(b):g}")
print(vm.pmf(0))
